pkg/pmem-csi-driver/nodeserver.go

```diff
@@ -89,29 +102,68 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	if len(req.GetTargetPath()) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
 	}
-	if len(req.GetStagingTargetPath()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Staging target path missing in request")
-	}
 
 	// Serialize by VolumeId
 	volumeMutex.LockKey(req.GetVolumeId())
 	defer volumeMutex.UnlockKey(req.GetVolumeId())
 
-	targetPath := req.TargetPath
-	sourcePath := req.StagingTargetPath
-	mounter := mount.New("")
+	var ephemeral bool
+	var device *pmdmanager.PmemDeviceInfo
+	var mountOptions []string
 
-	switch req.VolumeCapability.GetAccessType().(type) {
-	case *csi.VolumeCapability_Block:
-		dev, err := ns.dm.GetDevice(req.VolumeId)
+	srcPath := req.GetStagingTargetPath()
+	targetPath := req.GetTargetPath()
+
+	// Kubernetes v1.16+ would request ephemeral volumes via VolumeContext
+	val, ok := req.GetVolumeContext()[volumeContextEphemeral]
+	if ok {
+		b, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		ephemeral = b
+	} else {
+		// If met all below conditions we still treat as a ephemeral volume request:
+		// 1) No Volume found with given volume id
+		// 2) No provisioner info found in VolumeContext "storage.kubernetes.io/csiProvisionerIdentity"
+		// 3) No StagingPath in the request
+		device, err := ns.cs.dm.GetDevice(req.VolumeId)
+		if err != nil && err != os.ErrNotExist {
```

Comment:
Isn't `GetDevice` wrapping `os.ErrNoExist`? Then a direct comparison will always fails. You have to unwrap with `errors.Cause`, then compare.

Reply:
Yes, you are right, I should use `os.IsNotExist()`. This check was added when initially `GetDevice()` returning direct `os.ErrNotExist` without wrapping, later I added wrapping.